shared_model/utils/amount_utils.hpp

```diff
@@ -58,22 +58,50 @@ iroha::expected::Result<std::shared_ptr<shared_model::interface::Amount>,
                         std::shared_ptr<std::string>>
 operator-(const shared_model::interface::Amount &a,
           const shared_model::interface::Amount &b) {
-  // check precisions
-  if (a.precision() != b.precision()) {
-    return iroha::expected::makeError(
-        std::make_shared<std::string>("precision mismatch"));
-  }
   // check if a greater than b
   if (a.intValue() < b.intValue()) {
     return iroha::expected::makeError(
         std::make_shared<std::string>("minuend is smaller than subtrahend"));
   }
+  auto max_precision = std::max(a.precision(), b.precision());
+  auto val_a = a.intValue() * (int)std::pow(10, max_precision - a.precision());
+  auto val_b = b.intValue() * (int)std::pow(10, max_precision - b.precision());
+  if (val_a < a.intValue() || val_b < b.intValue()) {
+    return iroha::expected::makeError(
+        std::make_shared<std::string>("new precision overflows number"));
+  }
   return shared_model::builder::AmountBuilderWithoutValidator()
-      .precision(a.precision())
-      .intValue(a.intValue() - b.intValue())
+      .precision(max_precision)
+      .intValue(val_a - val_b)
       .build();
 }
 
+/**
+ * Make amount with bigger precision
+ * Otherwise nullopt is returned
+ * @param a amount
+ * @param b right term
+ */
+iroha::expected::Result<std::shared_ptr<shared_model::interface::Amount>,
+                        std::shared_ptr<std::string>>
+makePrecision(const shared_model::interface::Amount &amount,
+          const int new_precision) {
```

Comment:
do we need const here?

Reply:
Well, what is the problem with const here?